src/Parser/Monk/Brewmaster/Modules/Core/MasteryValue.js

```diff
@@ -0,0 +1,298 @@
+    if(spellId in this._hitCounts) {
+      return (this.dmg.byAbility(spellId).effective + this.dmg.staggeredByAbility(spellId)) / this._hitCounts[spellId];
+    }
+    return 0;
+  }
+
+  // events that either (a) add a stack or (b) can be dodged according
+  // to the data we have
+  get relevantTimeline() {
+    return this._timeline.filter(event => event.type === "cast" || this._dodgeableSpells[event.ability.guid]);
+  }
+
+  get _expectedValues() {
+    // expected damage mitigated according to the markov chain
+    let expectedDamageMitigated = 0;
+    let noMasteryExpectedDamageMitigated = 0;
+    // estimate of the damage that was actually dodged in this log
+    let estimatedDamageMitigated = 0;
+    // average dodge % across each event that could be dodged
+    let meanExpectedDodge = 0;
+    let dodgeableEvents = 0;
+
+    const stacks = new StackMarkovChain(); // mutating a const object irks me to no end
+    const noMasteryStacks = new StackMarkovChain();
+
+    // timeline replay is expensive, compute several things here and
+    // provide individual getters for each of the values
+    this.relevantTimeline.forEach(event => {
+      if(event.type === "cast") {
+        stacks.guaranteeStack();
+        noMasteryStacks.guaranteeStack();
+      } else if(event.type === "damage") {
+        const noMasteryDodgeChance = this.dodgeChance(noMasteryStacks.expected, 0, event.sourceID);
+        const expectedDodgeChance = this.dodgeChance(stacks.expected, event._masteryRating, event.sourceID);
+        const baseDodgeChance = this.dodgeChance(0, 0, event.sourceID);
+
+        const damage = (event.amount + event.absorbed) || this.meanHitByAbility(event.ability.guid);
+        expectedDamageMitigated += expectedDodgeChance * damage;
+        noMasteryExpectedDamageMitigated += noMasteryDodgeChance * damage;
+        estimatedDamageMitigated += (event.hitType === HIT_TYPES.DODGE) * damage;
+        meanExpectedDodge += expectedDodgeChance;
+        dodgeableEvents += 1;
+
+        stacks.processAttack(baseDodgeChance, this.stats.masteryPercentage(event._masteryRating, true));
+        noMasteryStacks.processAttack(baseDodgeChance, this.stats.masteryPercentage(0, true));
+      }
+    });
+
+    meanExpectedDodge /= dodgeableEvents;
+
+    return {
+      expectedDamageMitigated,
+      estimatedDamageMitigated,
+      meanExpectedDodge,
+      noMasteryExpectedDamageMitigated,
+    };
+  }
+
+  get expectedMitigation() {
+    return this._expectedValues.expectedDamageMitigated;
+  }
+  
+  get expectedMeanDodge() {
+    return this._expectedValues.meanExpectedDodge;
+  }
+
+  get actualDodgeRate() {
+    const totalHits = Object.keys(this._dodgeableSpells).reduce((sum, spellId) => sum + (this._hitCounts[spellId] || 0), 0);
+    const totalDodges = Object.keys(this._dodgeableSpells).reduce((sum, spellId) => sum + this._dodgeCounts[spellId], 0);
+    return totalDodges / (totalHits + totalDodges);
+  }
+
+  get estimatedActualMitigation() {
```

Comment:
I see the main metric you're using as a point of comparison between 'no mastery', 'expected', and 'actual' is Reduced DTPS. I understand that you need to use some number in order to produce an effective comparison, but I don't like Reduced DTPS.

Due to wildly differing tank damage done by bosses (and differing times that the Monk will actually have aggro), the Reduced DTPS numbers seem like they are only meaningful in relation to each other. There's also the inaccuracy added by occasionally having to guess the amount mitigated. Instead, why not just report the dodge rate, again comparing 'no mastery', 'expected', and 'actual'? It will be an easier number to read, and will require less guesswork.

Reply:
Dodge rate doesn't tell the whole story either. Some bosses have dangerous mechanics that can be dodged or have very hard-hitting melees, other bosses don't (see, for example, Starboi, on whom brm mastery does almost 0 mitigation).

You're right that DTPS only makes sense in comparison to another number. However, in this case we have another number available from the Damage Taken module. From the fight in the PR:

![screenshot from 2018-01-29 16-47-13](https://user-images.githubusercontent.com/4909458/35536190-24cf2016-0514-11e8-86d7-9f58f137a696.png)

so mastery reduced total DTPS by ~1/3 (disregarding all the complexity that stagger brings into that analysis).